### PR TITLE
feat: no duplicate for `fare_rules.txt` zone id fields

### LINKED ISSUE #356: Implement unique combination of fare_rules zone_id fields
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by lionel-nj**
closes #356

**Summary:**

This PR implements a new validation rule: the combination of `fare_rules.origin_id`, `fare_rules.contains_id` and `fare_rules.destination_id` fields should be unique in GTFS file `fare_rules.txt`.

**Expected behavior:** 

- a notice should be generated if two (or more rows) from GTFS files `fare_rules.txt` have duplicate value for the combination of  `fare_rules.origin_id`, `fare_rules.contains_id` and `fare_rules.destination_id` fields

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: -new feature short description-" (PR title must follow the [conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0/))
- [x] Linked all relevant issues
- ~[ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)~

### CODE REVIEW

**Comment by barbeau** [comment]
@lionel-nj I think fare_rules.route_id should be part of the uniqueness also? In other words you could have the same fare rule on two different routes?

@timMillet Could you please review the rule logic here to see if you agree?

**Comment by timMillet** [comment]
I agree with the rule logic: the combination of `fare_rules.route_id`, `fare_rules.origin_id`, `fare_rules.destination_id`, `fare_rules.contains_id` must be unique. Empty fields must be considered as part of the combination.

The rule is explicit in the spec so it should issue an error in the validator.

**Comment by lionel-nj** [comment]
Thanks @timMillet

File: RULES.md [hunk 1]
```diff
@@ -388,3 +388,9 @@ Per GTFS Best Practices, route alignments (in `shapes.txt`) should be within 100
 
 #### References:
 * [GTFS Best Practices shapes.txt](https://gtfs.org/best-practices/#shapestxt)
+
+<a name="DuplicateFareRuleZoneIdFieldsNotice"/>
+
+### DuplicateFareRuleZoneIdFieldsNotice
+
+The combination of `fare_rules.origin_id`, `fare_rules.contains_id` and `fare_rules.destination_id` fields should be unique in GTFS file `fare_rules.txt`.
```
Comment (barbeau) [inline RULES.md:396]: Needs to be updated to include route_id. Same with Javadocs

**Comment by lionel-nj** [comment]
@barbeau changes have been provided

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidator.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Override
+  public void validate(NoticeContainer noticeContainer) {
+    final Map<String, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
+        new HashMap<>(fareRuleTable.entityCount());
+    fareRuleTable
+        .getEntities()
+        .forEach(
+            fareRule -> {
+              if (fareRule.hasOriginId()
```
Comment (barbeau) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidator.java:50]: I don't think you should check for "hasX()" here? As Tim said uniqueness should be enforced on empty fields too.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        .setRouteId(routeId)
+        .setFareId(fareId)
+        .setOriginId(originId)
+        .setContainsId(containsId)
+        .setDestinationId(destinationId)
+        .build();
+  }
+
+  @Test
+  public void duplicateFareRuleZoneIdValuesShouldGenerateNotice() {
```
Comment (barbeau) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidatorTest.java:53]: Need test for some empty field values.

**Review by barbeau — COMMENTED** [review]
Thanks @lionel-nj, some comments inline.

**Comment by lionel-nj** [comment]
Also the rule logic has been reworked to enforce the uniqueness on empty fields values.

**Review by barbeau — APPROVED** [review]
👍
